Add defaults + split SignIn/SignOut

HaoK opened the pull request:
- Move SignIn/Signout into IAuthenticationHandlerSign[In\Out]Handler
- Add tests for AuthenticationService
- add overloads for SignOut/Forbid to use new default

@Tratcher

HaoK commented:
Fixes https://github.com/aspnet/Security/issues/1244 and https://github.com/aspnet/Security/issues/1264

Need to update Security/Hosts (to remove SignIn/Signout impls only) as a result of this PR as well

HaoK commented:
I'm going to combine the events changes we want for https://github.com/aspnet/Security/pull/1266 into this PR as well.  So the update to AuthenticationResult + rename of None => Ignore will go in this PR as well

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs:
```diff
@@ -93,7 +90,7 @@ public Task<AuthenticationScheme> GetDefaultForbidSchemeAsync()
         /// Returns the scheme that will be used by default for <see cref="IAuthenticationService.SignInAsync(HttpContext, string, System.Security.Claims.ClaimsPrincipal, AuthenticationProperties)"/>.
         /// This is typically specified via <see cref="AuthenticationOptions.DefaultSignInScheme"/>.
-        /// Otherwise, if only a single scheme exists, that will be used, if more than one exists, null will be returned.
+        /// Otherwise, this will fallback to <see cref="GetDefaultAuthenticateSchemeAsync"/>.
```
Update comment to include the single scenario?

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs:
```diff
@@ -121,6 +118,10 @@ public Task<AuthenticationScheme> GetDefaultSignOutSchemeAsync()
             {
                 return GetSchemeAsync(_options.DefaultSignOutScheme);
             }
+            if (_signOutHandlers.Count() == 1)
```
Add to doc comments?

campersau replied:
Why using Linq?

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs:
```diff
@@ -121,6 +118,10 @@ public Task<AuthenticationScheme> GetDefaultSignOutSchemeAsync()
+            {
+                return Task.FromResult(_signOutHandlers[0]);
+            }
             return GetDefaultSignInSchemeAsync();
```
There's now no guarantee that a signin handler supports signout. Maybe we should change the hierarchy to ISignIn -> ISignOut -> IAuth

HaoK replied:
I was just mirroring what we have in security today, where we have OIDC only implementing remote sign out...

HaoK replied:
Maybe we just tweak this also check that the handler for sign in scheme implements sign out before returning it, we can have hit continue to fallback to sign in

HaoK replied:
This problem also exists for the fallback for SignIn => Authenticate, so we need to add a similar check for that too

Tratcher replied:
No, the SignIn interface derives from the Auth interface, but SignOut does not derive from SignIn.

HaoK replied:
That's ok, I just mean:
```
 var signIn = GetDefaultSignInSchemeAsync();
 if (signIn != null && typeof(IAuthenticationSignOutHandler).IsAssignableFrom(signIn.HandlerType)) return signIn;
 else return null;
```

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs:
```diff
@@ -167,6 +168,14 @@ public void AddScheme(AuthenticationScheme scheme)
+                {
+                    _signOutHandlers.Add(scheme);
+                }
                 _map[scheme.Name] = scheme;
```
There's no check that the HandlerType actually implements IAuthenticationHandler

HaoK replied:
That's done in the scheme itself, https://github.com/aspnet/HttpAbstractions/blob/88a797cffddb43e586def38acc17c9435367a19a/src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationScheme.cs#L31

campersau commented on src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs:
```diff
@@ -102,11 +99,11 @@ public Task<AuthenticationScheme> GetDefaultSignInSchemeAsync()
                 return GetSchemeAsync(_options.DefaultSignInScheme);
             }
-            if (_map.Count == 1)
+            if (_signInHandlers.Count() == 1)
```
Why using Linq?

davidfowl commented on src/Microsoft.AspNetCore.Authentication.Abstractions/IAuthenticationSignOutHandler.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <summary>
+    /// Used to determine if a handler supports SignOut.
+    /// </summary>
+    public interface IAuthenticationSignOutHandler : IAuthenticationHandler
```
Why the derivation?

HaoK replied:
We require every handler to implement authenticate/challenge/forbid at a minimum, derivation makes this clearer, if we didn't require this, wouldn't that lead ppl to think they could only implement SignOut and use their handler?

davidfowl commented on src/Microsoft.AspNetCore.Authentication.Abstractions/IAuthenticationSignInHandler.cs:
```diff
@@ -0,0 +1,22 @@
+    /// <summary>
+    /// Used to determine if a handler supports SignIn.
+    /// </summary>
+    public interface IAuthenticationSignInHandler : IAuthenticationHandler
```
Why the derivation?

PinpointTownes commented:
So much sorcery in this PR... :cry: 

It would be much better to throw an `InvalidOperationException` when there's more than 1 scheme to encourage the developers to read the related documentation, so they can set the different schemes themselves.

PinpointTownes commented:
Related post: https://github.com/aspnet/Security/pull/1267#discussion_r122566994.

HaoK commented:
Updated the default sign/signout to enforce the handlers implement the interface, otherwise they return null (should this throw instead?)

HaoK commented:
We don't throw if the default scheme doesn't exist today, so it feels weird to throw

HaoK commented:
We'll also make SignIn derive from SignOut which dervies in Auth

HaoK commented:
Updated, SignInHandler is now the old uber AuthenticationHandler that has all methods implemented.  Renamed None() => NoResult() with the bool property Nothing => None (done for Security)

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs:
```diff
@@ -93,43 +90,49 @@ public Task<AuthenticationScheme> GetDefaultForbidSchemeAsync()
             if (_options.DefaultSignOutScheme != null)
             {
-                return GetSchemeAsync(_options.DefaultSignOutScheme);
+                return EnsureAssignable<IAuthenticationSignOutHandler>(await GetSchemeAsync(_options.DefaultSignOutScheme));
```
The type check is duplicated in SignOutAsync

HaoK replied:
Oh right, ok I can back out the extra complexity in the scheme provider then, and leave the type interface enforcement to the service

HaoK commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticateResult.cs:
```diff
@@ -64,22 +58,12 @@ public static AuthenticateResult Success(AuthenticationTicket ticket)
         /// </summary>
         /// <returns>The result.</returns>
-        public static AuthenticateResult None()
+        public static AuthenticateResult NoResult()
```
@Tratcher how about we keep this None, but call it NoResult() on the ResultContext's that will get exposed to events?

Tratcher replied:
That would leave you with a method and property that have the same name. NoResult is fine here.

HaoK replied:
These are currently called Nothing and None(), I just don't see much benefit renaming None => NoResult as we do have usages in the code internally already today (i.e. Identity), we can keep NoResult on ResultContext so the public API for events is still the same.

Tratcher replied:
NoResult() and None are better than None() and Nothing.

HaoK commented:
271faf11bbd5b05cd758f1c7e83eb59d45b6db59
